crossdock/server/server.py

```diff
@@ -130,79 +168,88 @@ def handle_trace_request(self, http_request, trace_request, span_handler, respon
+            trace_id = "%x" % span.trace_id
+            observed_span = ObservedSpan(
+                trace_id, span.is_sampled(),
+                span.get_baggage_item(constants.baggage_key))
 
-        if trace_request and trace_request.downstream is not None:
-            self.call_downstream(span, trace_request, tr, response_writer)
-        else:
-            response_writer.write(serializer.obj_to_json(tr))
-            response_writer.finish()
+            tr = TraceResponse(span=observed_span)
 
-    def call_downstream(self, span, trace_request, trace_response, response_writer):
-        if trace_request.downstream.transport == Transport.HTTP:
-            self.call_downstream_http(span, trace_request, trace_response, response_writer)
-        elif trace_request.downstream.transport == Transport.TCHANNEL:
-            if tchannel_supported:
-                self.real_call_downstream_tchannel(span, trace_request, trace_response,
-                                                   response_writer)
+            if trace_request and trace_request.downstream is not None:
+                call_downstream(span, trace_request, tr, response_writer)
             else:
-                self.call_downstream_tchannel(span, trace_request, trace_response,
-                                              response_writer)
-        else:
-            response_writer.finish()
-            raise UnknownTransportException("%s" % trace_request.downstream.transport)
-
-    def call_downstream_http(self, span, trace_request, trace_response, response_writer):
-
-        def handle_response(response):
-            tr = serializer.traceresponse_from_json(response.body)
-            if tr.notImplementedError:
                 response_writer.write(serializer.obj_to_json(tr))
-            else:
-                trace_response.downstream = tr
-                response_writer.write(serializer.obj_to_json(trace_response))
-            response_writer.finish()
-
-        downstream = trace_request.downstream
-        url = "http://%s:%s/join_trace" % (downstream.host, downstream.port)
-        body = serializer.downstream_to_json(downstream.downstream)
-
-        req = tornado.httpclient.HTTPRequest(url=url, method="POST",
-                                             headers={"Content-Type": "application/json"},
-                                             body=body)
-        http_client.before_http_request(tornado_http.TornadoRequestWrapper(request=req),
-                                        lambda: span)
-        client = tornado.httpclient.AsyncHTTPClient()
-        client.fetch(req, handle_response)
+                response_writer.finish()
 
-    def call_downstream_tchannel(self, span, trace_request, trace_response, response_writer):
 
-        tr = tchannelNotImplResponse()
-        response_writer.write(serializer.obj_to_json(tr))
+@tornado.gen.coroutine
+def call_downstream(span, trace_request, trace_response, response_writer):
```

Comment:
1. similar to above - we don't need `trace_request` here, only `downstream`
 1. let's have the functions without side effects - the should return downstream response, not modify the input parameter.